File: resilient/resilient-annotation-processor/src/main/java/ru/tinkoff/kora/resilient/annotation/processor/aop/CircuitBreakerKoraAspect.java
package ru.tinkoff.kora.resilient.annotation.processor.aop;

import com.squareup.javapoet.ClassName;
import com.squareup.javapoet.CodeBlock;
import ru.tinkoff.kora.annotation.processor.common.MethodUtils;
import ru.tinkoff.kora.annotation.processor.common.ProcessingError;
import ru.tinkoff.kora.annotation.processor.common.ProcessingErrorException;
import ru.tinkoff.kora.aop.annotation.processor.KoraAspect;

import javax.annotation.processing.ProcessingEnvironment;
import javax.lang.model.element.AnnotationMirror;
import javax.lang.model.element.ExecutableElement;
import javax.lang.model.type.DeclaredType;
import javax.lang.model.type.TypeMirror;
import javax.tools.Diagnostic;
import java.util.List;
import java.util.Optional;
import java.util.Set;
import java.util.concurrent.Future;

import static com.squareup.javapoet.CodeBlock.joining;

public class CircuitBreakerKoraAspect implements KoraAspect {

    private static final String ANNOTATION_TYPE = "ru.tinkoff.kora.resilient.circuitbreaker.annotation.CircuitBreaker";

    private final ProcessingEnvironment env;

    public CircuitBreakerKoraAspect(ProcessingEnvironment env) {
        this.env = env;
    }

    @Override
    public Set<String> getSupportedAnnotationTypes() {
        return Set.of(ANNOTATION_TYPE);
    }

    @Override
    public ApplyResult apply(ExecutableElement method, String superCall, AspectContext aspectContext) {
        if (MethodUtils.isFuture(method)) {
            throw new ProcessingErrorException("@CircuitBreaker can't be applied for types assignable from " + Future.class, method);
        }

        final Optional<? extends AnnotationMirror> mirror = method.getAnnotationMirrors().stream().filter(a -> a.getAnnotationType().toString().equals(ANNOTATION_TYPE)).findFirst();
        final String circuitBreakerName = mirror.flatMap(a -> a.getElementValues().entrySet().stream()
                .filter(e -> e.getKey().getSimpleName().contentEquals("value"))
                .map(e -> String.valueOf(e.getValue().getValue())).findFirst())
            .orElseThrow();

        var managerType = env.getTypeUtils().getDeclaredType(env.getElementUtils().getTypeElement("ru.tinkoff.kora.resilient.circuitbreaker.CircuitBreakerManager"));
        var fieldManager = aspectContext.fieldFactory().constructorParam(managerType, List.of());
        var circuitType = env.getTypeUtils().getDeclaredType(env.getElementUtils().getTypeElement("ru.tinkoff.kora.resilient.circuitbreaker.CircuitBreaker"));
        var fieldCircuit = aspectContext.fieldFactory().constructorInitialized(circuitType,
            CodeBlock.of("$L.get($S);", fieldManager, circuitBreakerName));

        final CodeBlock body;
        if (MethodUtils.isMono(method)) {
            body = buildBodyMono(method, superCall, fieldCircuit);
        } else if (MethodUtils.isFlux(method)) {
            body = buildBodyFlux(method, superCall, fieldCircuit);
        } else {
            body = buildBodySync(method, superCall, fieldCircuit);
        }

        return new ApplyResult.MethodBody(body);
    }

    private CodeBlock buildBodySync(ExecutableElement method, String superCall, String circuitBreakerField) {
        final CodeBlock superMethod = buildMethodCall(method, superCall);
        final CodeBlock methodCall = MethodUtils.isVoid(method)
            ? superMethod
            : CodeBlock.of("var t = $L", superMethod.toString());

        final CodeBlock returnCall = MethodUtils.isVoid(method)
            ? CodeBlock.of("return")
            : CodeBlock.of("return t", superMethod.toString());

        final ClassName cbException = ClassName.get("ru.tinkoff.kora.resilient.timeout", "TimeoutExhaustedException");

        return CodeBlock.builder().add("""
            var _circuitBreaker = $L;
            try {
                _circuitBreaker.acquire();
                $L;
                _circuitBreaker.releaseOnSuccess();
                $L;
            } catch ($T e) {
                throw e;
            } catch (Exception e) {
                _circuitBreaker.releaseOnError(e);
                throw e;
            }
            """, circuitBreakerField, methodCall.toString(), returnCall.toString(), cbException).build();
    }

    private CodeBlock buildBodyMono(ExecutableElement method, String superCall, String circuitBreakerField) {
        final CodeBlock superMethod = buildMethodCall(method, superCall);

        return CodeBlock.builder().add("""
            var _circuitBreaker = $L;
                        
            return Mono.defer(() -> {
                  _circuitBreaker.acquire();
                  return $L
                      .doOnSuccess(r -> _circuitBreaker.releaseOnSuccess())
                      .doOnError(_circuitBreaker::releaseOnError);
            });
                 """, circuitBreakerField, superMethod.toString()).build();
    }

    private CodeBlock buildBodyFlux(ExecutableElement method, String superCall, String circuitBreakerField) {
        final CodeBlock superMethod = buildMethodCall(method, superCall);

        return CodeBlock.builder().add("""
            var _circuitBreaker = $L;
                        
            return Flux.defer(() -> {
                  _circuitBreaker.acquire();
                  return $L
                      .doOnComplete(() -> _circuitBreaker.releaseOnSuccess())
                      .doOnError(_circuitBreaker::releaseOnError);
            });
                 """, circuitBreakerField, superMethod.toString()).build();
    }

    private CodeBlock buildMethodCall(ExecutableElement method, String call) {
        return method.getParameters().stream().map(p -> CodeBlock.of("$L", p)).collect(joining(", ", call + "(", ")"));
    }
}

File: resilient/resilient-symbol-processor/src/main/kotlin/ru/tinkoff/kora/resilient/symbol/processor/aop/CircuitBreakerKoraAspect.kt
package ru.tinkoff.kora.resilient.symbol.processor.aop

import com.google.devtools.ksp.KspExperimental
import com.google.devtools.ksp.getClassDeclarationByName
import com.google.devtools.ksp.processing.Resolver
import com.google.devtools.ksp.symbol.KSFunctionDeclaration
import com.squareup.kotlinpoet.CodeBlock
import com.squareup.kotlinpoet.MemberName
import com.squareup.kotlinpoet.ksp.toClassName
import reactor.core.publisher.Flux
import reactor.core.publisher.Mono
import ru.tinkoff.kora.aop.symbol.processor.KoraAspect
import ru.tinkoff.kora.ksp.common.FunctionUtils.isFlow
import ru.tinkoff.kora.ksp.common.FunctionUtils.isFlux
import ru.tinkoff.kora.ksp.common.FunctionUtils.isFuture
import ru.tinkoff.kora.ksp.common.FunctionUtils.isMono
import ru.tinkoff.kora.ksp.common.FunctionUtils.isSuspend
import ru.tinkoff.kora.ksp.common.FunctionUtils.isVoid
import ru.tinkoff.kora.ksp.common.exception.ProcessingError
import ru.tinkoff.kora.ksp.common.exception.ProcessingErrorException
import java.util.concurrent.Future
import javax.tools.Diagnostic

@KspExperimental
class CircuitBreakerKoraAspect(val resolver: Resolver) : KoraAspect {

    companion object {
        const val ANNOTATION_TYPE: String = "ru.tinkoff.kora.resilient.kora.CircuitBreaker"
    }

    override fun getSupportedAnnotationTypes(): Set<String> {
        return setOf(ANNOTATION_TYPE)
    }

    override fun apply(method: KSFunctionDeclaration, superCall: String, aspectContext: KoraAspect.AspectContext): KoraAspect.ApplyResult {
        if (method.isFuture()) {
            throw ProcessingErrorException("@CircuitBreaker can't be applied for types assignable from ${Future::class.java}", method)
        } else if (method.isMono()) {
            throw ProcessingErrorException("@CircuitBreaker can't be applied for types assignable from ${Mono::class.java}", method)
        } else if (method.isFlux()) {
            throw ProcessingErrorException("@CircuitBreaker can't be applied for types assignable from ${Flux::class.java}", method)
        }

        val annotation = method.annotations.filter { a -> a.annotationType.resolve().toClassName().canonicalName == ANNOTATION_TYPE }.first()
        val circuitBreakerName = annotation.arguments.asSequence().filter { arg -> arg.name!!.getShortName() == "value" }.map { arg -> arg.value.toString() }.first()

        val managerType = resolver.getClassDeclarationByName("ru.tinkoff.kora.resilient.kora.circuitbreaker.CircuitBreakerManager")!!.asType(listOf())
        val fieldManager = aspectContext.fieldFactory.constructorParam(managerType, listOf())
        val circuitType = resolver.getClassDeclarationByName("ru.tinkoff.kora.resilient.kora.circuitbreaker.CircuitBreaker")!!.asType(listOf())
        val fieldCircuit = aspectContext.fieldFactory.constructorInitialized(
            circuitType,
            CodeBlock.of("%L[%S]", fieldManager, circuitBreakerName)
        )

        val body = if (method.isFlow()) {
            buildBodyFlow(method, superCall, fieldCircuit)
        } else if (method.isSuspend()) {
            buildBodySync(method, superCall, fieldCircuit)
        } else {
            buildBodySync(method, superCall, fieldCircuit)
        }

        return KoraAspect.ApplyResult.MethodBody(body)
    }

    private fun buildBodySync(
        method: KSFunctionDeclaration, superCall: String, fieldCircuitBreaker: String
    ): CodeBlock {
        val superMethod = buildMethodCall(method, superCall)
        val methodCall = if(method.isVoid()) superMethod else CodeBlock.of("val t = %L", superMethod)
        val returnCall = if(method.isVoid()) CodeBlock.of("") else CodeBlock.of("t")

        return CodeBlock.builder().add(
            """
            return try {
                %L.acquire()
                %L
                %L.releaseOnSuccess()
                %L
            } catch (e: ru.tinkoff.kora.resilient.kora.circuitbreaker.CircuitBreakerNotPermittedException) {
                throw e
            } catch (e: java.lang.Exception) {
                %L.releaseOnError(e)
                throw e
            }
            """.trimIndent(), fieldCircuitBreaker, methodCall, fieldCircuitBreaker, returnCall, fieldCircuitBreaker
        ).build()
    }

    private fun buildBodyFlow(
        method: KSFunctionDeclaration, superCall: String, fieldCircuitBreaker: String
    ): CodeBlock {
        val flowMember = MemberName("kotlinx.coroutines.flow", "flow")
        val emitMember = MemberName("kotlinx.coroutines.flow", "emitAll")
        val superMethod = buildMethodCall(method, superCall)
        return CodeBlock.builder().add(
            """
            return %M {
                try {
                    %L.acquire()
                    %M(%L)
                    %L.releaseOnSuccess()
                } catch (e: ru.tinkoff.kora.resilient.kora.circuitbreaker.CircuitBreakerNotPermittedException) {
                    throw e
                } catch (e: java.lang.Exception) {
                    %L.releaseOnError(e)
                    throw e
                }
            }
            """.trimIndent(), flowMember, fieldCircuitBreaker, emitMember, superMethod.toString(), fieldCircuitBreaker, fieldCircuitBreaker
        ).build()
    }

    private fun buildMethodCall(method: KSFunctionDeclaration, call: String): CodeBlock {
        return CodeBlock.of(method.parameters.asSequence().map { p -> CodeBlock.of("%L", p) }.joinToString(", ", "$call(", ")"))
    }
}
